From 294aff13a417ee9ad60422c4a7712e8766d2d6ec Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Sun, 8 Nov 2015 17:21:01 +0200
Subject: [PATCH 0472/1240] armada3700: cci-400: Enable snoop and dvm on port
 S3

- Enable Snoop and DVM on port S3 (SPL only)

Change-Id: Id050c7d1625a7229b723d45a61e6d424259370a6
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/24660
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 arch/arm/include/asm/arch-armadalp/regs-base.h |  4 ++++
 drivers/misc/mvebu_io_addr_dec.c               | 13 ++++++++++++-
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/arch/arm/include/asm/arch-armadalp/regs-base.h b/arch/arm/include/asm/arch-armadalp/regs-base.h
index bccacba..46eeffd 100644
--- a/arch/arm/include/asm/arch-armadalp/regs-base.h
+++ b/arch/arm/include/asm/arch-armadalp/regs-base.h
@@ -39,4 +39,8 @@
 #define MVEBU_ARLP_GBE0_INTERNAL_REG_BASE	MVEBU_REGS_BASE + 0xc364
 #define MVEBU_ARLP_GBE1_INTERNAL_REG_BASE	MVEBU_REGS_BASE + 0xc464
 
+/* CCI-400 */
+#define MVEBU_CCI_BASE			(MVEBU_REGS_BASE + 0x8000000)
+#define MVEBU_CCI_S3_SNOOP_CTRL_REG	(MVEBU_CCI_BASE + 0x4000)
+
 #endif	/* _REGS_BASE_H_ */
diff --git a/drivers/misc/mvebu_io_addr_dec.c b/drivers/misc/mvebu_io_addr_dec.c
index be9e005..6ca13e1 100644
--- a/drivers/misc/mvebu_io_addr_dec.c
+++ b/drivers/misc/mvebu_io_addr_dec.c
@@ -16,8 +16,8 @@
  * ***************************************************************************
  */
 
-#define DEBUG
 #include <common.h>
+#include <asm/io.h>
 #include <asm/arch-mvebu/mvebu.h>
 
 int init_a3700_io_addr_dec(void)
@@ -26,6 +26,17 @@ int init_a3700_io_addr_dec(void)
 
 	debug_enter();
 
+	/*
+	   CCI-400 enable snoop and dvm on S3 port.
+	   For details see the <CoreLink CCI-400 Cache Coherent Interconnect> document.
+	   bit[0] - Enable issuing of snoop requests from this slave interface.
+	   bit[1] - Enable issuing of DVM message requests from this slave interface
+	   bit[29:2] - Reserved
+	   bit[30] - Slave interface supports snoops
+	   bit[31] - Slave interface supports DVM messages
+	 */
+	writel(0xC0000003, MVEBU_CCI_S3_SNOOP_CTRL_REG);
+
 	/* Add units configuration code here */
 
 	debug_exit();
-- 
1.9.1

